File: src/main/java/com/example/crudapi/controller/AuthController.java
package com.example.crudapi.controller;

import com.example.crudapi.dto.VerifyOtpRequest;
import com.example.crudapi.dto.VerifyOtpResponse;
import com.example.crudapi.model.User;
import com.example.crudapi.repository.UserRepository;
import com.google.firebase.FirebaseApp;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseAuthException;
import com.google.firebase.auth.FirebaseToken;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Base64;
import java.util.UUID;

@RestController
@RequestMapping("/api/auth")
@CrossOrigin(origins = "*")
public class AuthController {

    private static final String DEFAULT_ROLE = "EMPLOYEE";

    @Autowired
    private UserRepository userRepository;

    @PostMapping("/verify-otp")
    public ResponseEntity<VerifyOtpResponse> verifyOtp(@Valid @RequestBody VerifyOtpRequest request) {
        if (FirebaseApp.getApps().isEmpty()) {
            return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE)
                    .body(new VerifyOtpResponse(null, null, null,
                            "Firebase not configured (set FIREBASE_CREDENTIALS_PATH)"));
        }

        FirebaseToken decoded;
        try {
            decoded = FirebaseAuth.getInstance().verifyIdToken(request.getIdToken());
        } catch (FirebaseAuthException e) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(new VerifyOtpResponse(null, null, null, "Invalid or expired Firebase token"));
        }

        Object phoneClaim = decoded.getClaims().get("phone_number");
        if (!(phoneClaim instanceof String) || ((String) phoneClaim).isBlank()) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(new VerifyOtpResponse(null, null, null, "Token has no phone_number"));
        }
        String phone = (String) phoneClaim;

        User user = userRepository.findByPhone(phone).orElseGet(() -> {
            User u = new User();
            u.setPhone(phone);
            u.setRole(DEFAULT_ROLE);
            return u;
        });
        userRepository.save(user);

        String token = Base64.getEncoder()
                .encodeToString((user.getPhone() + ":" + UUID.randomUUID()).getBytes());

        return ResponseEntity.ok(new VerifyOtpResponse(
                user.getPhone(), user.getRole(), token, "Login successful"));
    }
}

File: node/src/auth/auth.controller.ts
import {
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Inject,
  Post,
  Req,
  Res,
  UseGuards,
} from '@nestjs/common';
import { Response } from 'express';
import { randomUUID } from 'crypto';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { ConfigService } from '@nestjs/config';
import type * as admin from 'firebase-admin';
import { User } from './user.entity';
import { VerifyOtpRequest } from './dto/verify-otp-request.dto';
import { VerifyOtpResponse } from './dto/verify-otp-response.dto';
import { AuthGuard, AuthedRequest } from './auth.guard';
import { FIREBASE_AUTH, FirebaseAuthService } from '../firebase/firebase.module';

@Controller('api/auth')
export class AuthController {
  constructor(
    @InjectRepository(User)
    private readonly users: Repository<User>,
    @Inject(FIREBASE_AUTH)
    private readonly firebaseAuth: FirebaseAuthService,
    private readonly config: ConfigService,
  ) {}

  @Post('verify-otp')
  @HttpCode(HttpStatus.OK)
  async verifyOtp(
    @Body() body: VerifyOtpRequest,
    @Res({ passthrough: true }) res: Response,
  ): Promise<VerifyOtpResponse> {
    if (!this.firebaseAuth) {
      res.status(HttpStatus.SERVICE_UNAVAILABLE);
      return new VerifyOtpResponse(
        null as unknown as string,
        null,
        null,
        'Firebase not configured (set FIREBASE_CREDENTIALS_PATH)',
      );
    }

    let decoded: admin.auth.DecodedIdToken;
    try {
      decoded = await this.firebaseAuth.verifyIdToken(body.idToken);
    } catch {
      res.status(HttpStatus.UNAUTHORIZED);
      return new VerifyOtpResponse(
        null as unknown as string,
        null,
        null,
        'Invalid or expired Firebase token',
      );
    }

    const phone = decoded.phone_number;
    if (!phone) {
      res.status(HttpStatus.UNAUTHORIZED);
      return new VerifyOtpResponse(
        null as unknown as string,
        null,
        null,
        'Token has no phone_number',
      );
    }

    const adminPhone = (
      this.config.get<string>('BOOTSTRAP_ADMIN_PHONE', '') ?? ''
    ).trim();

    let user = await this.users.findOne({ where: { phone } });
    if (!user) {
      user = this.users.create({ phone, role: null });
    }
    if (adminPhone && phone === adminPhone) {
      user.role = 'ADMIN';
    }
    await this.users.save(user);

    const token = Buffer.from(`${user.phone}:${randomUUID()}`).toString('base64');

    return new VerifyOtpResponse(
      user.phone,
      user.role ?? null,
      token,
      'Login successful',
    );
  }

  @Get('me')
  @UseGuards(AuthGuard)
  me(@Req() req: AuthedRequest) {
    const u = req.user!;
    return { id: u.id, phone: u.phone, role: u.role ?? null };
  }
}
